File: Data Structures & Algorithms/binary-tree-right-side-view/submission-1.java
class Solution {
    public List<Integer> rightSideView(TreeNode root) {
        TreeNode cur = root;
        List<Integer> list = new LinkedList<>();
        Stack<Pair> stack = new Stack<>();
        int curDepth = 0;
        int maxDepth = 0;

        while (!stack.isEmpty() || cur != null) {
            if (cur == null) {
                Pair curPair = stack.pop();
                curDepth = curPair.depth;
                cur = curPair.node;
            }

            if (cur.left != null) stack.add(new Pair(cur.left, curDepth + 1));

            if (curDepth == maxDepth) {
                list.add(cur.val);
                maxDepth++;
            }

            curDepth++;
            cur = cur.right;
        }

        return list;
    }
}

class Pair {
    Pair(TreeNode node, int depth) {
        this.node = node; this.depth = depth;
    }

    TreeNode node;
    int depth;
}

File: Data Structures & Algorithms/same-binary-tree/submission-1.cpp
class Solution {
public:
    bool same = true;

    bool isSameTree(TreeNode* p, TreeNode* q) {
      helper(p, q);
      return same;
    }

    void helper(TreeNode* p, TreeNode* q) {
      if (p == nullptr and q == nullptr) return;
      else if (p == nullptr) goto notSame;
      else if (q == nullptr) goto notSame;
      else if (p->val != q->val) goto notSame;

      helper(p->left, q->left);
      helper(p->right, q->right);
      return;

      notSame:
        same = false;
        return;
    }
};
